src/kbc_modules/online_sev_kbc/mod.rs

```diff
@@ -0,0 +1,209 @@
+            .arg(SECRET_MODULE_NAME)
+            .status()?
+            .success()
+        {
+            return Err(anyhow!("Failed to load secret module."));
+        }
+        Ok(SecretKernelModule {})
+    }
+}
+impl Drop for SecretKernelModule {
+    fn drop(&mut self) {
+        Command::new(MODPROBE_PATH)
+            .arg("-r")
+            .arg(SECRET_MODULE_NAME)
+            .status()
+            .expect("Failed to unload secret module.");
+    }
+}
+
+pub struct OnlineSevKbc {
+    // KBS info for compatibility; unused
+    kbs_info: HashMap<String, String>,
+
+    // Known ciphers, corresponding to wrap_type
+    ciphers: Ciphers,
+    kbs_uri: String,
+    connection: Result<Connection>,
+}
+
+fn get_ciphers() -> Ciphers {
+    // The sample KBC uses aes-gcm (Rust implementation). The offline file system KBC uses OpenSSL
+    // instead to get access to hardware acceleration on more platforms (e.g. s390x). As opposed
+    // to aes-gcm, OpenSSL will only allow GCM when using AEAD. Because authentication is not
+    // handled here, AEAD cannot be used, therefore, CTR is used instead.
+    [(String::from("aes_256_ctr"), Cipher::aes_256_ctr())]
+        .iter()
+        .cloned()
+        .collect()
+}
+
+#[async_trait]
+impl KbcInterface for OnlineSevKbc {
+    fn check(&self) -> Result<KbcCheckInfo> {
+        Ok(KbcCheckInfo {
+            kbs_info: self.kbs_info.clone(),
+        })
+    }
+
+    async fn decrypt_payload(&mut self, annotation: &str) -> Result<Vec<u8>> {
+        let annotation_packet: AnnotationPacket = serde_json::from_str(annotation)
+            .map_err(|e| anyhow!("Failed to parse annotation: {}", e))?;
+
+        let key = self.get_key_from_kbs(annotation_packet.kid).await?;
+
+        let iv = decode(annotation_packet.iv).map_err(|e| anyhow!("Failed to decode IV: {}", e))?;
+        let wrapped_data = decode(annotation_packet.wrapped_data)
+            .map_err(|e| anyhow!("Failed to decode wrapped key: {}", e))?;
+        let wrap_type = annotation_packet.wrap_type;
+
+        let cipher = self
+            .ciphers
+            .get(&wrap_type)
+            .ok_or_else(|| anyhow!("Received unknown wrap type: {}", wrap_type))?;
+        // Redact decryption errors to avoid oracles
+        decrypt(*cipher, &key, Some(&iv), &wrapped_data).map_err(|_| anyhow!("Failed to decrypt"))
+    }
+}
+
+impl OnlineSevKbc {
+    #[allow(clippy::new_without_default)]
+    pub fn new(kbs_uri: String) -> OnlineSevKbc {
+        OnlineSevKbc {
+            kbs_info: HashMap::new(),
+            ciphers: get_ciphers(),
+            kbs_uri,
+            connection: load_connection(),
+        }
+    }
+
+    async fn query_kbs(&self, secret_type: String, secret_id: String) -> Result<Vec<u8>> {
+        let uri = format!("http://{}", self.kbs_uri).parse::<Uri>()?;
+
+        let channel = tonic::transport::Channel::builder(uri).connect_lazy();
+        let mut client = KeyBrokerServiceClient::new(channel);
```

Comment:
(future) Consider setting-up channel+client in `new()` and re-using them in each request, this might improve performance if you need to fetch several secrets in a short time. Though we'll need to see how this works when parallel requests are made.